### ISSUE TITLE
False positive in manual_map with unsafe fn

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#7820]
```rust
pub trait OptionExt<T> {
    #[allow(clippy::missing_safety_doc)]
    unsafe fn unsafe_map<U>(self, op: unsafe fn(T) -> U) -> Option<U>;
}

impl<T> OptionExt<T> for Option<T> {
    unsafe fn unsafe_map<U>(self, op: unsafe fn(T) -> U) -> Option<U> {
        match self {
            Some(t) => Some(op(t)),
            None => None,
        }
    }
}
```

```console
$ cargo clippy -- -Dclippy::all

error: manual implementation of `Option::map`
  --> src/main.rs:8:9
   |
8  | /         match self {
9  | |             Some(t) => Some(op(t)),
10 | |             None => None,
11 | |         }
   | |_________^ help: try this: `self.map(op)`
   |
   = note: `-D clippy::manual-map` implied by `-D clippy::all`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_map
```

Clippy's suggested replacement does not compile.

```console
error[E0277]: expected a `FnOnce<(T,)>` closure, found `unsafe fn(T) -> U`
 --> src/main.rs:8:18
  |
8 |         self.map(op)
  |              --- ^^ expected an `FnOnce<(T,)>` closure, found `unsafe fn(T) -> U`
  |              |
  |              required by a bound introduced by this call
  |
  = help: the trait `FnOnce<(T,)>` is not implemented for `unsafe fn(T) -> U`
```

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.57.0-nightly (dfc5add91 2021-10-13)
binary: rustc
commit-hash: dfc5add915e8bf4accbb7cf4de00351a7c6126a1
commit-date: 2021-10-13
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```
